File: src/Main/java/com/nishant/algorithms/Sorting/QuickSort/QuickSort.java
package com.nishant.algorithms.Sorting.QuickSort;

public class QuickSort {

    public static void sort(int[] arr) {
        sort(arr, 0, arr.length);
    }

    private static void sort(int[] arr, int left, int right) {
        if(left >= right - 1) {
            return;
        }

        int pivotIndex = partition(arr, left, right);
        sort(arr, left, pivotIndex);
        sort(arr, pivotIndex + 1, right);
    }

    private static int partition(int[] arr, int left, int right) {
        int pivotIndex = right - 1, partitionIndex = left;

        for(int i = left; i < right - 1; i ++) {
            if(arr[i] < arr[pivotIndex]) {
                if(partitionIndex == pivotIndex) {
                    pivotIndex = i;
                }

                swap(arr, i, partitionIndex ++);
            }
        }

        swap(arr, pivotIndex, partitionIndex);
        return partitionIndex;
    }

    private static void swap(int[] arr, int pos1, int pos2) {
        int temp = arr[pos1];
        arr[pos1] = arr[pos2];
        arr[pos2] = temp;
    }

}

File: src/main/python/sorting/quicksort/quick_sort.py
def quick_sort(to_sort):
    quick_sort_helper(to_sort, 0, len(to_sort))


def quick_sort_helper(to_sort, left, right):
    if left >= right - 1:
        return

    partition_index = partition(to_sort, left, right)
    quick_sort_helper(to_sort, left, partition_index)
    quick_sort_helper(to_sort, partition_index + 1, right)


def partition(to_sort, left, right):
    pivot = to_sort[right - 1]
    partition_index = left

    for i in range(left, right - 1, 1):
        if to_sort[i] < pivot:
            to_sort[partition_index], to_sort[i] = to_sort[i], to_sort[partition_index]
            partition_index += 1

    to_sort[right - 1], to_sort[partition_index] = to_sort[partition_index], to_sort[right - 1]
    return partition_index


if __name__ == "__main__":
    arr = [-2, -1, 0, 0]
    print(str(arr))
    quick_sort(arr)
    print(str(arr))
